File: JAVA/20_Divide&Conquer/BOJ_02_1992_쿼드트리.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class Main {

    static int[][] arr;
    static StringBuilder sb = new StringBuilder();

    public static boolean compressd(int x, int y, int size) {
        int color = arr[x][y];
        for (int i = x; i < x + size; i++) {
            for (int j = y; j < y + size; j++) {
                if (arr[i][j] != color) {
                    return false;
                }
            }
        }
        return true;
    }

    public static void quadTree(int x, int y, int size) {
        if (compressd(x, y, size)) {
            sb.append(arr[x][y]);
            return;
        }
        sb.append("(");
        size /= 2;
        for (int i = 0; i < 2; i++) {
            for (int j = 0; j < 2; j++) {
                quadTree(x + i * size, y + j * size, size);
            }
        }
        sb.append(")");
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        arr = new int[n][n];
        for (int i = 0; i < n; i++) {
            String str = br.readLine();
            for (int j = 0; j < n; j++) {
                arr[i][j] = str.charAt(j) - '0';
            }
        }
        quadTree(0, 0, n);
        System.out.println(sb);
    }
}

File: C/20_Divide&Conquer/BOJ_02_1992_쿼드트리.c
#include <stdio.h>
int arr[64][64];
int compressed(int x, int y, int size) {
	int color = arr[x][y];
	for (int i = x; i < x + size; i++) {
		for (int j = y; j < y + size; j++) {
			if (arr[i][j] != color) {
				return 0;
			}
		}
	}
	return 1;
}
void quadTree(int x, int y, int size) {
	if (compressed(x, y, size)) {
		printf("%d", arr[x][y]);
		return;
	}
	printf("(");
	size /= 2;
	for (int i = 0; i < 2; i++) {
		for (int j = 0; j < 2; j++) {
			quadTree(x + i * size, y + j * size, size);
		}
	}
	printf(")");
}
main() {
	int n;
	scanf("%d", &n);
	char* str = (char*)malloc(sizeof(char) * n);
	for (int i = 0; i < n; i++) {
		scanf("%s", str);
		for (int j = 0; j < n; j++) {
			arr[i][j] = str[j] - '0';
		}
	}
	quadTree(0, 0, n);
}